Allow formEncode/urlEncode to infer @safe

ichordev opened the pull request:
All `formEncodeImpl` overloads now check for the presence of `T.byKeyValue`, and prefer using it over `opApply`, thereby making `formEncode`/`urlEncode` infer `@safe` in the majority of cases.

I also added some extra unittests, and corrected a comment mentioning a 'json map' when the example instead used associative array.

Fixes #16.

Commit:
Allow `formEncode`/`urlEncode` to infer safe

ichordev commented:
Not again! 😭 I’ll fix this later today.

s-ludwig commented:
Have you already had a look whether your editor can support EditorConfig? I have .editorconfig files in all of my projects, as that makes it a lot easier to avoid at least the most trivial formatting issues (indentation, newlines, encoding), when working across different projects. Just for the record (I never used dfmt myself apart from some tests) - I've also added some dfmt specific directives and added a K&R style mode to [dfmt](https://github.com/dlang-community/dfmt) to get reasonably close to the [style guide](https://vibed.org/style-guide) (which is not enforced).

ichordev commented:
> Have you already had a look whether your editor can support EditorConfig? I have .editorconfig files in all of my projects, as that makes it a lot easier to avoid at least the most trivial formatting issues (indentation, newlines, encoding), when working across different projects.

Ah, I need a plugin for that. Thank you for letting me know.

> Just for the record (I never used dfmt myself apart from some tests) - I've also added some dfmt specific directives and added a K&R style mode to [dfmt](https://github.com/dlang-community/dfmt) to get reasonably close to the [style guide](https://vibed.org/style-guide) (which is not enforced).

I'll try this next time if the plugin doesn't work. :)

For now, I believe I have gotten rid of the heretical whitespace!

s-ludwig commented on source/vibe/inet/webform.d:
```diff
@@ -453,23 +469,51 @@ unittest {
+		string r = value.type == T.Type.string ? value.get!string : value.to!string;
+} ());
+
+private enum isJsonLikeWithByKeyValue(T) = __traits(compiles, () {
+	import std.conv;
+	foreach (string key, T value; T.init.byKeyValue())
+		string r = value.type == T.Type.string ? value.get!string : value.to!string;
+} ());
```
```suggestion
private enum isJsonLikeWithByKeyValue(T) = __traits(compiles, () {
	import std.conv;
	foreach (item; T.init.byKeyValue())
		string k = item.key, r = item.value.type == T.Type.string ? item.value.get!string : item.value.to!string;
} ());
```

For some reason the latest DMD hangs when using the separate key/value foreach syntax. Changing this here and also at the two places where the actual iteration happens fixed this for me.

ichordev replied:
Oh neat! If I’d known that was valid I would’ve used that all along.
Unfortunately I’m away from my computer until tomorrow but if you want to push to my fork to expedite things then feel free. Otherwise I can do it sometime tomorrow.

ichordev replied:
Done!

ichordev commented:
5 minutes?! Did dub respond really slowly, or did I actually make it compile that much slower? :’)

Commit:
Fix DMD hang: switch to 1-item foreach syntax for Json
